internal/app/backend/backend_service.go

```diff
@@ -180,9 +182,12 @@ func synchronizeRecv(ctx context.Context, syncStream synchronizerStream, m *sync
 }
 
 // callMmf triggers execution of MMFs to fetch match proposals.
-func callMmf(ctx context.Context, cc *rpc.ClientCache, req *pb.FetchMatchesRequest, proposals chan<- *pb.Match) error {
+func callMmf(ctx context.Context, cc *rpc.ClientCache, req *pb.FetchMatchesRequest, functionConnectionTimeout time.Duration, proposals chan<- *pb.Match) error {
 	defer close(proposals)
 	address := fmt.Sprintf("%s:%d", req.GetConfig().GetHost(), req.GetConfig().GetPort())
+	if err := isAddressAvailable(address, functionConnectionTimeout); err != nil {
```

Comment:
This feels like a hack/bandaid.  
I've found that gRPC is weird and operates with different primitives than people tend to be used to with http requests, however they're very well thought out, and will have a correct way to accomplish things.

Specifically, gRPC client doesn't necessarily represent a specific tcp connection, instead it may be connected to multiple  instances and go between working and broken.  What we're saying here is "if this connection isn't currently alive, we want to fail faster than we otherwise would".  gRPC has a couple constructs which would be useful here:

https://pkg.go.dev/google.golang.org/grpc#ClientConn.WaitForStateChange
https://pkg.go.dev/google.golang.org/grpc#ClientConn.GetState

To implement this change, I think we should solve it at a system level instead of just the MMF (fixing this problem with evaluator too).  So that means changing the client cache: https://github.com/googleforgames/open-match/blob/master/internal/rpc/clientcache.go

Something like,
```golang
func (cc *ClientCache) GetGRPC(address string) (*grpc.ClientConn, error) {
	val, exists := cc.cache.Load(address)
	c, ok := val.(cachedGRPCClient)
	if !ok || !exists {
		conn, err := GRPCClientFromEndpoint(cc.cfg, address)
		if err != nil {
			return nil, err
		}
		c = cachedGRPCClient{
			client: conn,
		}
		cc.cache.Store(address, c)
	}
 
        timeoutDuration := readConnectionReadyTimeout()
        ctx := context.WithTimeout(context.Background(), timeoutDuration)
        for {
            s := c.client.GetState()
            if s == connectivity.Ready {
                break
            }
            if !c.client.WaitForStateChange(ctx, s) {
              return nil, status.Errorf(codes.Unavailable, "timeout waiting for connection ready after %s, stuck in state %s", timeoutDuration, s)
            }
         }

	return c.client, nil
}
```

There might be an even better way to specify this, but these apis are complex so I'm not sure :/
Maybe ask on the gRPC slack?

(also see if it's reasonable to pass the context from the call here, don't use it for dialing but only this connection's specific waiting.)

Also, please write a test covering this behavior.

Reply:
Thanks for the feedback, it felt a bit hacky originally but I'm still a bit new to the gRPC primitives so I wasn't sure what I could use.

Reply:
I mostly looked at https://pkg.go.dev/google.golang.org/grpc to find something which would do what we want.